### PR TITLE
[Merged by Bors] - hare: fix memory leak

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

fix hare memory leak #3991

## Changes

- change hare time-related configuration to using Duration to facilitate testing
- create a test that fails under old code (stuck at `consensusProcess.terminate()`)
- separate cleanup and terminate logic
  - move waiting for errgroup into a separate function `Stop()`
  - keep created consensus process inside hare, and invoke `Stop()` after unregistering from message broker

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

File: hare/algorithm_test.go [hunk 1]
```diff
@@ -182,48 +178,59 @@ func (mev *mockEligibilityValidator) ValidateEligibilityGossip(context.Context,
...
+	mo.EXPECT().Proof(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).AnyTimes()
+	mo.EXPECT().CalcEligibility(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), proc.nid, gomock.Any()).Return(uint16(1), nil).AnyTimes()
+	proc.oracle = mo
+
+	proc.value = NewSetFromValues(types.ProposalID{1}, types.ProposalID{2})
+	proc.Start()
+	require.Eventually(t, func() bool {
+		proc.Stop()
+		return true
+	}, 500*time.Millisecond, 100*time.Millisecond)
```
Comment (poszu) [inline hare/algorithm_test.go:234]: It could be simplified to:
```suggestion
	proc.Stop()
```

The test will timeout if `proc.Stop()` never returns.

File: hare/hare.go [hunk 2]
```diff
@@ -174,8 +173,8 @@ func (h *Hare) HandleEligibility(ctx context.Context, emsg *types.HareEligibilit
 }
 
 func (h *Hare) getLastLayer() types.LayerID {
-	h.layerLock.RLock()
-	defer h.layerLock.RUnlock()
+	h.mu.Lock()
+	defer h.mu.Unlock()
```
Comment (poszu) [inline hare/hare.go:210]: Why did you change the `RWMutex` to a `Mutex`? It made sense to lock for reading only here.

File: hare/hare.go [hunk 3]
```diff
@@ -206,9 +205,15 @@ func (h *Hare) outOfBufferRange(id types.LayerID) bool {
 }
 
 func (h *Hare) oldestResultInBuffer() types.LayerID {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.oldestResultInBufferLocked()
```
Comment (poszu) [inline hare/hare.go:243]: It could benefit from an `RWMutex` too

File: hare/hare.go [hunk 4]
```diff
@@ -342,14 +344,43 @@ func (h *Hare) onTick(ctx context.Context, id types.LayerID) (bool, error) {
...
+	defer h.mu.Unlock()
+	h.cps[cp.ID()] = cp
+	logger.With().Info("number of consensus processes (after register)",
+		log.Int("count", len(h.cps)))
+}
+
+func (h *Hare) getCP(lid types.LayerID) Consensus {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.cps[lid]
```
Comment (poszu) [inline hare/hare.go:403]: It could benefit from `RWMutex` too if the `value, ok := h.cps[lid]` notation was used.

**Review by poszu — APPROVED** [review]
LGTM.

I'd bring back the `RWMutex` though.

File: hare/algorithm_test.go (hunk 1 above)
Comment (countvonzero) [reply hare/algorithm_test.go:234]: i don't see why test timing out is better. when i do `go test` it will just hang.

File: hare/hare.go (hunk 2 above)
Comment (countvonzero) [reply hare/hare.go:210]: mostly because i prefer code simplicity over premature optimization. 

- Mutex is more straightforward to use
- RWMutex.Lock is slower than Mutex.Lock
- in hare module, the only place one needs to worry about thread-safety is the gossip handler. the rest should not have lock contention

i'd use RWMutex only when the component is write-once and read-many-times. otherwise i prefer simplicity.

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge
